resources/user.go

```diff
@@ -0,0 +1,212 @@
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package resources
+
+import (
+	"fmt"
+	errwrap "github.com/pkg/errors"
+	"log"
+	"os/user"
+)
+
+func init() {
+	RegisterResource("user", func() Res { return &UserRes{} })
+}
+
+//TODO
+type UserRes struct {
+	BaseRes `yaml:",inline"`
+	Gid     string `yaml:"gid"`
+	HomeDir string `yaml:"homedir"`
+	State   string `yaml:"state"` // state: exists, absent
+	Uid     string `yaml:"uid"`
+	exec    *ExecRes
+}
+
+// Default returns some sensible defaults for this resource.
+func (obj *UserRes) Default() Res {
+	return &UserRes{
+		BaseRes: BaseRes{
+			MetaParams: DefaultMetaParams, // force a default
+		},
+	}
+}
+
+// Validate if the params passed in are valid data.
+func (obj *UserRes) Validate() error {
+	if obj.State != "exists" && obj.State != "absent" {
+		return fmt.Errorf("State must be 'exists' or 'absent'")
+	}
+	return obj.BaseRes.Validate()
+}
+
+// Init initializes the resource.
+func (obj *UserRes) Init() error {
+	obj.exec = &ExecRes{}
+	obj.exec.Init()
+	return obj.BaseRes.Init() // call base init, b/c we're overriding
+}
+
+// Watch is the primary listener for this resource and it outputs events.
+func (obj *UserRes) Watch() error {
+
+	// notify engine that we're running
+	if err := obj.Running(); err != nil {
+		return err // bubble up a NACK...
+	}
+
+	var send = false // send event?
+	var exit *error
+	for {
+		select {
+		// NOTE: this part is very similar to the file resource code
+		case event := <-obj.Events():
+			// we avoid sending events on unpause
+			if exit, send = obj.ReadEvent(event); exit != nil {
+				return *exit // exit
+			}
+		}
+
+		// do all our event sending all together to avoid duplicate msgs
+		if send {
+			send = false
+			obj.Event()
+		}
+	}
+}
+
+// CheckApply method for User resource. Does nothing, returns happy!
+func (obj *UserRes) CheckApply(apply bool) (checkOK bool, err error) {
+	log.Printf("%s: CheckApply(%t)", obj, apply)
+
+	var exists = true
+
+	usr, err := user.Lookup(obj.GetName())
+
+	log.Printf("%+v", usr)
+
+	if err != nil {
+		log.Printf("the user: %s does not exist", obj.GetName())
+		exists = false
+	}
+
+	if obj.State == "exists" && exists &&
+		(obj.Gid == "" || obj.Gid == usr.Gid) &&
+		(obj.HomeDir == "" || obj.HomeDir == usr.HomeDir) &&
+		(obj.Uid == "" || obj.Uid == usr.Uid) {
+		return true, nil
+	}
+
+	if obj.State == "absent" && !exists {
+		return true, nil
+	}
+
+	if !apply {
+		return false, nil
+	}
+
+	var cmd string
+	if obj.State == "exists" {
+		if !exists {
+			cmd = "useradd "
+		}
+		if exists {
+			cmd = "usermod "
+		}
+		cmd += obj.GetName()
+		if obj.HomeDir != "" {
+			cmd += " -d " + obj.HomeDir
+		}
+		if obj.Gid != "" {
+			cmd += " -g " + obj.Gid
+		}
+		if obj.Uid != "" {
+			cmd += " -u " + obj.Uid
+		}
+		obj.exec.Cmd = cmd
+	}
+
+	if obj.State == "absent" {
+		obj.exec.Cmd = "userdel " + obj.GetName()
+	}
+
+	if _, err := obj.exec.CheckApply(apply); err != nil {
```

Comment:
why drop the first return value? You'll need that to know what to return if it's not an error.

Reply:
Why/where do I need to return anything if it's not an error?